### PR TITLE
Create clean run of CircleCI (linting fixes)

### PR DESCRIPTION
**Pull request by jfrerich**
#### Summary
Fix Linting errors for a clean CI build

### CODE REVIEW

**Comment by hanzei** [comment]
@hanzei Please ping me when this is ready for review

**Comment by jfrerich** [comment]
@hanzei and @mickmister, this PR doesn't solve all of the linting issue, but does resolve a majority of them.  I'd love to get this merged soon so that I can incorporate the changes in https://github.com/mattermost/mattermost-app-zendesk/pull/33

Thanks!

File: src/forms/subscriptions.ts [hunk 1]
```diff
@@ -1,18 +1,21 @@
...
-import {AppSelectOption, AppCall, AppForm, AppField} from 'mattermost-redux/types/apps';
+import {AppSelectOption, AppCallRequest, AppForm, AppField} from 'mattermost-redux/types/apps';
+import {AppContextWithBot} from 'types/apps';
 import {AppFieldTypes} from 'mattermost-redux/constants/apps';
 import Client4 from 'mattermost-redux/client/client4.js';
 
 import {newZDClient, newMMClient, ZDClient} from '../clients';
 import {getStaticURL, Routes} from '../utils';
-import {getBulletedList, makeSubscriptionOptions, makeChannelOptions, getChannelIDFromTriggerTitle, tryPromiseWithMessage} from '../utils/utils';
+import {getBulletedList, makeSubscriptionOptions, makeChannelOptions, getChannelIDFromTriggerTitle, tryPromiseWithMessage, ZDFieldOption} from '../utils/utils';
```
Comment (mickmister) [inline src/forms/subscriptions.ts:9]: Should `getBulletedList` be named `makeBulletedList`?

File: src/forms/subscriptions.ts [hunk 2]
```diff
@@ -1,18 +1,21 @@
...
+import {ZDTrigger, ZDTriggerCondition} from '../utils/ZDTypes';
 import {SubscriptionFields, ZendeskIcon} from '../utils/constants';
 import {BaseFormFields} from '../utils/base_form_fields';
 
 // newSubscriptionsForm returns a form response to create subscriptions
-export async function newSubscriptionsForm(call: AppCall): Promise<AppForm> {
-    const zdClient = await newZDClient(call.context);
-    const mmClient = newMMClient(call.context).asAdmin();
+export async function newSubscriptionsForm(call: AppCallRequest): Promise<AppForm> {
+    const context: AppContextWithBot = call.context as AppContextWithBot;
```
Comment (mickmister) [inline src/forms/subscriptions.ts:16]: Either declare the type or use `as`. No need to use both

File: src/forms/subscriptions.ts [hunk 3]
```diff
@@ -362,12 +363,12 @@ class FormFields extends BaseFormFields {
         return this.builder.getFieldLabelByName(SubscriptionFields.SubSelectName);
     }
 
-    getSubTriggerByID(subID: string): any {
-        const triggers = this.getChannelTriggers(this.getSelectedChannelID());
-        return triggers.find((trigger) => String(trigger.id) === subID);
+    getSubTriggerByID(subID: string): ZDTrigger {
+        const triggers: ZDTrigger[] = this.getChannelTriggers(this.getSelectedChannelID());
+        return triggers.find((trigger) => String(trigger.id) === subID) as ZDTrigger;
```
Comment (mickmister) [inline src/forms/subscriptions.ts:368]: Are we guaranteed to find a trigger with this id here?

File: src/restapi/fIncomingWebhooks.ts [hunk 4]
```diff
@@ -16,30 +18,31 @@ export async function fHandleSubcribeNotification(req: Request, res: Response):
 
     // TODO: we need zendesk bot admin so that admin requests can be made by the
     // bot and not from an actiing user
-    const fakeBotContext: AppContext = {
+    const fakeBotContext: AppContextWithBot = {
+        app_id: getManifest().app_id,
         acting_user_id: 'rgixs6uimp88tq8x8w3yxu3oqe',
-    };
+    } as AppContextWithBot;
```
Comment (mickmister) [inline src/restapi/fIncomingWebhooks.ts:24]: There's no bot id in this context. Why is it being casted to `AppContextWithBot`?

**Review by mickmister — CHANGES_REQUESTED** [review]
Ideally there would be less `as` going on, but the code does look a bit more defensive now. I have some questions and a few requests

File: .eslintrc.json [hunk 5]
```diff
@@ -63,7 +63,13 @@
...
       2,
       {
-        "properties": "never"
+        "properties": "never",
+        "allow": [
+          "bot_access_token", 
+          "acting_user_id", 
+          "acting_user_access_token",
+          "ticket_field_ids"
+        ]
```
[the remark this answers is not in the document]
Comment (hanzei) [reply .eslintrc.json:72]: Does JSON.Parse handle snake_case vs CamelCase correctly? I.e. can we abstract the JSON format away?

File: src/app/app.ts [hunk 6]
```diff
@@ -24,14 +26,14 @@ export interface App {
...
+    context: AppContextWithBot
     values: AppCallValues
 
-    constructor(call: AppCall) {
+    constructor(call: AppCallRequest) {
         this.call = call;
-        this.context = call.context;
-        this.values = call.values;
+        this.context = call.context as AppContextWithBot;
+        this.values = call.values as AppCallValues;
```
Comment (hanzei) [inline src/app/app.ts:36]: Is the `as` really needed?

File: src/app/app.ts [hunk 7]
```diff
@@ -24,14 +26,14 @@ export interface App {
...
+    call: AppCallRequest
+    context: AppContextWithBot
     values: AppCallValues
 
-    constructor(call: AppCall) {
+    constructor(call: AppCallRequest) {
         this.call = call;
-        this.context = call.context;
-        this.values = call.values;
+        this.context = call.context as AppContextWithBot;
```
Comment (hanzei) [inline src/app/app.ts:35]: Not every `Call` has an actin user. How do we deal with that case?

File: src/app/app.ts [hunk 8]
```diff
@@ -120,11 +122,11 @@ class AppImpl implements App {
...
 
         // add bot to team and channel
-        const botUserID = this.context.bot_user_id;
-        const addToTeamReq = adminClient.addToTeam(this.context.team_id, botUserID);
+        const botUserID = this.context.bot_user_id as string;
+        const addToTeamReq = adminClient.addToTeam(this.context.team_id as string, botUserID);
         await tryPromiseWithMessage(addToTeamReq, 'Failed to add bot to team');
 
-        const addToChannelReq = adminClient.addToChannel(botUserID, this.context.channel_id);
+        const addToChannelReq = adminClient.addToChannel(botUserID, this.context.channel_id as string);
```
Comment (hanzei) [inline src/app/app.ts:129]: Why are these `as`'s needed? In which cases are the values `undefined`?

File: src/app/triggerFromForm.ts [hunk 9]
```diff
@@ -9,14 +11,14 @@ interface TriggerFromFrom {
...
-    values: AppFormValues;
+    values: AppCallValues;
     context: AppContext;
-    trigger: any
+    trigger: ZDTrigger
 
-    constructor(call: AppCall) {
-        this.values = call.values as AppFormValues;
+    constructor(call: AppCallRequest) {
+        this.values = call.values as AppCallValues;
```
Comment (hanzei) [inline src/app/triggerFromForm.ts:19]: Why is this needed?

File: src/app/triggerFromForm.ts [hunk 10]
```diff
@@ -9,14 +11,14 @@ interface TriggerFromFrom {
...
         this.context = call.context;
-        this.trigger = {};
+        this.trigger = {} as ZDTrigger;
```
Comment (hanzei) [inline src/app/triggerFromForm.ts:21]: Does it make sense to use `?ZDTrigger` as the type for `trigger` instead?

File: src/forms/create_ticket.ts [hunk 11]
```diff
@@ -64,17 +68,17 @@ class FormFields extends BaseFormFields {
...
     }
 
     async addFormSelectDependentFields(): Promise<void> {
         // get the form id from the selected form field value
-        const formID = this.builder.getFieldValueByName(CreateTicketFields.NameFormsSelect);
+        const formID = this.builder.getFieldValueByName(CreateTicketFields.NameFormsSelect) as unknown;
 
-        const zdFormFieldIDs = this.getTicketFieldIDs(formID);
-        const fieldsListReq = this.zdClient.ticketfields.list();
+        const zdFormFieldIDs = this.getTicketFieldIDs(formID as number);
```
Comment (hanzei) [inline src/forms/create_ticket.ts:80]: Why not use `ToString` instead of two type assertions?

File: src/forms/create_ticket.ts [hunk 12]
```diff
@@ -154,7 +160,7 @@ class FormFields extends BaseFormFields {
             case ZDFieldTypes.Muliselect: {
                 f.type = AppFieldTypes.STATIC_SELECT;
                 const options = this.isSystemField(field) ? field.system_field_options : field.custom_field_options;
-                f.options = makeOptions(options);
+                f.options = makeOptions(options as ZDFieldOption[]);
```
Comment (hanzei) [inline src/forms/create_ticket.ts:163]: Why does `makeOptions` not return an `ZDFieldOption[]`?

File: src/forms/create_ticket.ts [hunk 13]
```diff
@@ -171,7 +177,7 @@ class FormFields extends BaseFormFields {
     }
 
     isSystemField(field: Users.Fields.UserField): boolean {
-        return (SystemFields.includes(String(field.type)) || field.system_field_options);
+        return (SystemFields.includes(String(field.type)) || field.system_field_options) as boolean;
```
Comment (hanzei) [inline src/forms/create_ticket.ts:180]: TS reports:
```
Property 'system_field_options' does not exist on type 'UserField'. Did you mean 'custom_field_options'?ts(2551)
```

File: src/forms/subscriptions.ts [hunk 14]
```diff
@@ -222,17 +223,17 @@ class FormFields extends BaseFormFields {
...
         }
         return [];
     }
 
     // getAnyConditions returns an array of Zendesk ALL trigger conditions
-    getAllConditions(): any[] {
+    getAllConditions(): ZDTriggerCondition[] {
         if (this.getSelectedSubTrigger() && this.getSelectedSubTrigger().conditions) {
-            return this.getSelectedSubTrigger().conditions.all;
+            return this.getSelectedSubTrigger().conditions.all as ZDTriggerCondition[];
```
Comment (hanzei) [inline src/forms/subscriptions.ts:236]: Why shouldn't the caller deal with the case that `conditions.all` is `undefined´?

File: src/forms/zendesk_config.ts [hunk 15]
```diff
@@ -32,8 +34,8 @@ class FormFields extends BaseFormFields {
     configStore: ConfigStore
     storeValues: AppConfigStore
 
-    constructor(call: AppCall, configStore: ConfigStore, zdClient: ZDClient, mmClient: Client4) {
-        super(call, zdClient, mmClient);
+    constructor(call: AppCallRequest, configStore: ConfigStore, mmClient: Client4) {
+        super(call, {} as ZDClient, mmClient);
```
Comment (hanzei) [inline src/forms/zendesk_config.ts:38]: `BaseFormFields` doesn't require a ´ZDClient`. Can we pass `undefined` here?

File: src/restapi/fComplete.ts [hunk 16]
```diff
@@ -53,11 +53,12 @@ export async function fComplete(req: Request, res: Response): Promise<void> {
...
 // access to the PluginKVStore and retrieves the token and url from state
-function createContextFromState(botToken: string, url: string): AppContext {
-    const context: AppContext = {
+function createContextFromState(botToken: string, url: string): AppContextWithBot {
+    const context: AppContextWithBot = {
+        app_id: '',
         bot_access_token: botToken,
         mattermost_site_url: url,
-    };
+    } as AppContextWithBot;
```
Comment (hanzei) [inline src/restapi/fComplete.ts:61]: Do we need the `as`?

File: src/restapi/fConnect.ts [hunk 17]
```diff
@@ -1,13 +1,15 @@
...
 import {newOKCallResponseWithMarkdown} from '../utils/call_responses';
 
 import {newConfigStore} from '../store';
 
-import {Routes, createOAuthState, contextFromRequest} from '../utils';
+import {Routes, createOAuthState} from '../utils';
 
 export async function fConnect(req: Request, res: Response): Promise<void> {
-    const context = contextFromRequest(req);
+    const context: AppContextWithBot = req.body.context;
```
Comment (hanzei) [inline src/restapi/fConnect.ts:12]: 1/5 that I like the helper here.

File: src/restapi/fDisconnect.ts [hunk 18]
```diff
@@ -27,7 +29,7 @@ export async function fDisconnect(req: Request, res: Response): Promise<void> {
 
     // delete the token from the store
     const tokenStore = newTokenStore(context);
-    const deleteTokenReq = tokenStore.deleteToken(context.acting_user_id);
+    const deleteTokenReq = tokenStore.deleteToken(context.acting_user_id as string);
```
Comment (hanzei) [inline src/restapi/fDisconnect.ts:32]: The types should make it clear that there is an `actiong_user_id`. Do we need a new type?

File: src/restapi/fIncomingWebhooks.ts (hunk 4 above)
Comment (hanzei) [reply src/restapi/fIncomingWebhooks.ts:24]: Actually, there should be a bot that we can use, right? See L31.

File: src/restapi/fIncomingWebhooks.ts [hunk 19]
```diff
@@ -1,9 +1,11 @@
 import {Post} from 'mattermost-redux/types/posts';
 
 import {Request, Response} from 'express';
+import {AppContextWithBot} from 'types/apps';
 import {AppContext} from 'mattermost-redux/types/apps';
```
Comment (hanzei) [inline src/restapi/fIncomingWebhooks.ts:5]: Is this needed any longer?

File: src/store/oauth.ts [hunk 20]
```diff
@@ -31,7 +31,7 @@ class TokenStoreImpl implements TokenStore {
 }
 
 export const newTokenStore = (context: AppContext): TokenStore => {
```
Comment (hanzei) [inline src/store/oauth.ts:33]: Should this be 
```suggestion
export const newTokenStore = (context: AppContextWithBot): TokenStore => {
```

**Review by hanzei — CHANGES_REQUESTED** [review]

File: src/app/app.ts [hunk 21]
```diff
@@ -120,11 +122,11 @@ class AppImpl implements App {
         const adminClient = newMMClient(this.context).asActingUser();
 
         // add bot to team and channel
-        const botUserID = this.context.bot_user_id;
-        const addToTeamReq = adminClient.addToTeam(this.context.team_id, botUserID);
+        const botUserID = this.context.bot_user_id as string;
```
[the remark this answers is not in the document]
Comment (jfrerich) [reply src/app/app.ts:125]: This struct says it is optional

https://github.com/mattermost/mattermost-plugin-apps/blob/master/apps/context.go#L27

File: src/app/app.ts (hunk 8 above)
Comment (jfrerich) [reply src/app/app.ts:129]: @hanzei, the values require the `as` type casting when the type field is optional.  https://github.com/mattermost/mattermost-plugin-apps/blob/master/apps/context.go#L27

Comment (jfrerich) [reply src/app/app.ts:129]: @mickmister, maybe the structs tags are not correct and have `omitempty` when that are actually required?

File: src/restapi/fComplete.ts [hunk 22]
```diff
@@ -53,11 +53,12 @@ export async function fComplete(req: Request, res: Response): Promise<void> {
 
 // createContextFromState constructs an AppContext from state. fComplete needs
 // access to the PluginKVStore and retrieves the token and url from state
-function createContextFromState(botToken: string, url: string): AppContext {
-    const context: AppContext = {
+function createContextFromState(botToken: string, url: string): AppContextWithBot {
+    const context: AppContextWithBot = {
```
[the remark this answers is not in the document]
Comment (jfrerich) [reply src/restapi/fComplete.ts:57]: done.

File: src/restapi/fComplete.ts (hunk 16 above)
Comment (jfrerich) [reply src/restapi/fComplete.ts:61]: Yes. it is needed.  Without, it, the variable `context` will be an object of unknown type

File: src/store/oauth.ts (hunk 20 above)
Comment (jfrerich) [reply src/store/oauth.ts:33]: yes. thanks and done.

File: src/forms/subscriptions.ts (hunk 1 above)
Comment (jfrerich) [reply src/forms/subscriptions.ts]: done.

File: src/forms/subscriptions.ts (hunk 2 above)
Comment (jfrerich) [reply src/forms/subscriptions.ts:18]: done.

File: src/forms/create_ticket.ts [hunk 23]
```diff
@@ -35,9 +39,9 @@ export async function newCreateTicketForm(call: AppCall): Promise<AppForm> {
 class FormFields extends BaseFormFields {
     zdTicketForms: ZDFormFieldOption[]
     postMessage: string
-    constructor(call: AppCall, zdClient: ZDClient, mmClient: Client4) {
+    constructor(call: AppCallRequest, zdClient: ZDClient, mmClient: Client4) {
         super(call, zdClient, mmClient);
-        this.postMessage = call.context.post.message;
+        this.postMessage = call?.context?.post?.message as string;
```
[the remark this answers is not in the document]
Comment (hmhealey) [reply src/forms/create_ticket.ts:44]: I'd think the first CallWithContext/CallWithPost part should work, but I can't think of if there's any way to shorthand that. I don't think there's a good generic way to tell TypeScript "yeah, that says optional, but I know it's set" since that sounds like what you're asking about though

File: src/restapi/fConnect.ts (hunk 17 above)
Comment (jfrerich) [reply src/restapi/fConnect.ts:12]: I don't mind adding this helper func back, but it will need to change.  The AppContext will change depending on the expand values.  I think I'd need to add another input that defined the `type` of context that is expected and what should be returned from the function.

Or.. I could create several helpers that accept a `req`, but return and App Context that is appropriate.  For example 
`contextWithBotFromRequest(req)`
`contextWithPostFromRequest(req)`

Comment (jfrerich) [reply src/restapi/fConnect.ts:12]: By not having the helpers, we can just cast the type when defining the variable.

`const context: AppContextWithBot: = req.body.context`

File: src/app/app.ts (hunk 6 above)
Comment (jfrerich) [reply src/app/app.ts:36]: @hanzei, yes, `values` field in the `CallRequest` struct is an optional field as seen 

https://github.com/mattermost/mattermost-plugin-apps/blob/master/apps/call.go#L55

and 

https://github.com/mattermost/mattermost-redux/blob/master/src/types/apps.ts#L71

File: src/app/triggerFromForm.ts (hunk 9 above)
Comment (jfrerich) [reply src/app/triggerFromForm.ts:19]: Unfortunately. See comment https://github.com/mattermost/mattermost-app-zendesk/pull/34/files#r612100603

I suppose I could create `AppCallRequestWithValues` type that has them as a required field. What do you think?

File: src/app/triggerFromForm.ts [hunk 24]
```diff
@@ -9,14 +11,14 @@ interface TriggerFromFrom {
...
     }
 
     getTrigger(): any {
```
[the remark this answers is not in the document]
Comment (jfrerich) [reply src/app/triggerFromForm.ts]: Totally makes sense.  It is not defined by the node-zendesk type definitions.  I created a new type and use it now.

[the remark this answers is not in the document]
Comment (jfrerich) [reply src/app/triggerFromForm.ts]: See commit [`0856c9a` (#34)](https://github.com/mattermost/mattermost-app-zendesk/pull/34/commits/0856c9a66430876490353dec65d824f926f7977e)

File: src/forms/create_ticket.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (jfrerich) [reply src/forms/create_ticket.ts:187]: done.

File: src/forms/create_ticket.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (jfrerich) [reply src/forms/create_ticket.ts]: @mickmister, @hmhealey please see the latest changes to this file and the creation of the following file.

https://github.com/mattermost/mattermost-app-zendesk/blob/lint-fixes/src/types/apps.ts

File: src/forms/create_ticket.ts (hunk 12 above)
Comment (jfrerich) [reply src/forms/create_ticket.ts:171]: Essentially, this function is converting a `Zendesk` Form Field to an `Apps` `AppField` 

`makeOptions` accepts `ZDFieldOption[]` and returns options for an `AppField`.  Does that make sense?

File: src/utils/ZDTypes.ts [hunk 25]
```diff
@@ -0,0 +1,20 @@
...
+
+export type ZDTriggerCondition = {
+    field: string;
+    operator: string;
+    value?: string;
+}
+
+export type ZDTriggerConditions = {
+    all?: ZDTriggerCondition[];
+    any: ZDTriggerCondition[];
```
[the remark this answers is not in the document]
Comment (jfrerich) [reply src/utils/ZDTypes.ts:11]: In the context of how the App currently operates, this type is only used with subscriptions and currently only supports the `any` field.  I added the `all` optional field because it will eventually be supported by the app.

File: src/app/triggerFromForm.ts (hunk 10 above)
Comment (jfrerich) [reply src/app/triggerFromForm.ts:19]: @hanzei, I'm not quite sure I understand your suggestion. I understand the ability to define a field as optional, using the `?` prefix, but I don't know how to use that in the context of defining an empty object for `ZDTrigger`

File: src/app/app.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (hanzei) [reply src/app/app.ts:125]: It's not optional.

File: src/app/app.ts [hunk 26]
```diff
@@ -120,21 +122,21 @@ class AppImpl implements App {
...
         const botClient = newMMClient(this.context).asBot();
 
-        const post: Post = {
+        const post = {
             message,
             user_id: botUserID,
             channel_id: String(this.context.channel_id),
             root_id: String(this.context.post_id),
-        };
+        } as Post;
```
Comment (hanzei) [inline src/app/app.ts:139]: Is here a good use case for ´Partial`?

File: src/app/app.ts (hunk 8 above)
Comment (hanzei) [reply src/app/app.ts]: The `omitempty` is correct. But in all context's that `createBotPost` is called, there will be a `channel_id`.

File: src/app/triggerFromForm.ts (hunk 10 above)
Comment (hanzei) [reply src/app/triggerFromForm.ts:19]: After reading the class a second time I think it should be restructured. There is separate discussion about of the object oriented pattern is really needed, but I'm not going into that.

My suggestion is to the generation of the trigger into the constructor and letting `getTrigger` just returned the class field `trigger`.

File: src/forms/create_ticket.ts (hunk 12 above)
Comment (hanzei) [reply src/forms/create_ticket.ts:171]: I've misread the code.

Still, I think there is some case handling missing there. If `options` is `undefined` that line with throw an error because `field.system_field_options` might be `undefined`. Can that case happen?

File: src/forms/create_ticket.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (hmhealey) [reply src/forms/create_ticket.ts]: That works for me.

The other thing I could think to do for that would be to just have two objects passed around: the AppContext and some generic Context that could be modified depending on caller. That simplifies typing, but obviously makes it more annoying to work with since you're passing two objects around instead of once.

File: src/restapi/fIncomingWebhooks.ts (hunk 4 above)
Comment (jfrerich) [reply src/restapi/fIncomingWebhooks.ts]: A refactor has removed the need for this fake context variable

File: src/app/app.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (jfrerich) [reply src/app/app.ts:169]: fixed.

File: src/forms/create_ticket.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (jfrerich) [reply src/forms/create_ticket.ts]: To solve this problem, I created some concrete Ctx* types that are used when the request is read.  Any function calls that were using AppContext were changed to access a finite number of inputs. (mattermost_site_url, bot_access_token, .etc)

File: src/app/app.ts (hunk 7 above)
Comment (jfrerich) [reply src/app/app.ts]: I created multiple concreate types with required fields and the request in that call function determines which to use.

File: src/forms/create_ticket.ts [hunk 27]
```diff
@@ -64,17 +68,17 @@ class FormFields extends BaseFormFields {
 
     // setState sets state for the current Zendesk Form
     async setState(): Promise<void> {
-        const listReq = this.zdClient.ticketforms.list();
+        const listReq = this.zdClient?.ticketforms.list();
```
[the remark this answers is not in the document]
Comment (jfrerich) [reply src/forms/create_ticket.ts:87]: Yes, you are correct. Thanks.

File: src/forms/create_ticket.ts (hunk 13 above)
Comment (jfrerich) [reply src/forms/create_ticket.ts]: I created a custom ZDField type to handle this case.  https://github.com/mattermost/mattermost-app-zendesk/pull/34/files#diff-317fa1c3e53f7eddff384b174597cc031681277978d4f4ce2e88f13c0834fa5fR29

File: src/forms/zendesk_config.ts (hunk 15 above)
Comment (jfrerich) [reply src/forms/zendesk_config.ts:48]: yes, I had to allow `undefined` in `.eslintrc.json`.   done.

File: src/restapi/fConnect.ts (hunk 17 above)
Comment (jfrerich) [reply src/restapi/fConnect.ts]: I went with creating concrete types that are now used by the call functions.  I could still create the helper functions which would be good for checking the existence of the values.

File: src/restapi/fDisconnect.ts (hunk 18 above)
Comment (jfrerich) [reply src/restapi/fDisconnect.ts]: The token store has since been removed from this PR when Oauth handling was added to the proxy app.  What this question does bring up is the possible need to expose a proxy plugin endpoint to revoke/delete a token

File: src/restapi/fIncomingWebhooks.ts (hunk 19 above)
Comment (jfrerich) [reply src/restapi/fIncomingWebhooks.ts]: There are multiple new custom context types that are used by the call functions.

File: .eslintrc.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (jfrerich) [reply .eslintrc.json]: FYI, I disable this check in one file where it was causing issues (warnings)

File: src/utils/utils.ts [hunk 28]
```diff
@@ -22,17 +23,20 @@ type ZDSubscriptionFieldOption = {
...
 // trigger title
 export const getChannelIDFromTriggerTitle = (title: string): string => {
     const re = new RegExp(SubscriptionFields.RegexTriggerTitle);
-    const match = title.match(re);
+    const match = title.match(re) || '';
     if (!match) {
-        console.log('malformed Mattermost Trigger title', match[1]);
+        console.log('malformed Mattermost Trigger title', title);
     }
     return match[1];
```
[the remark this answers is not in the document]
Comment (jfrerich) [reply src/utils/utils.ts]: @mickmister, is it appropriate to throw and error instead so the function exists?

```js
throw new Error('Failed to get user access_token' + title);
```

File: src/utils/utils.ts [hunk 29]
```diff
@@ -22,17 +23,20 @@ type ZDSubscriptionFieldOption = {
 
 const getDisplaySubTitleOption = (option: ZDSubscriptionFieldOption): string => {
     const re = new RegExp(SubscriptionFields.RegexTriggerTitle);
-    const newTitle = option.title.match(re)[2];
-    return newTitle;
+    const newTitle = option.title.match(re) || '';
+    if (!newTitle) {
+        console.log('malformed Mattermost Trigger title', newTitle);
+    }
+    return newTitle[2];
```
[the remark this answers is not in the document]
Comment (jfrerich) [reply src/utils/utils.ts]: converted to throw an error

File: src/forms/subscriptions.ts (hunk 3 above)
Comment (jfrerich) [reply src/forms/subscriptions.ts:395]: converted to throw an error

File: src/app/triggerFromForm.ts (hunk 10 above)
Comment (jfrerich) [reply src/app/triggerFromForm.ts:21]: done.

File: src/app/app.ts (hunk 8 above)
Comment (jfrerich) [reply src/app/app.ts]: This code has been removed. The bot can post to any channel now and does not need to be added to a channel or team

File: src/forms/create_ticket.ts (hunk 11 above)
Comment (jfrerich) [reply src/forms/create_ticket.ts]: remove the `as unknown` and now casting formID as a string.

File: src/forms/subscriptions.ts (hunk 14 above)
Comment (jfrerich) [reply src/forms/subscriptions.ts]: done.

File: src/forms/create_ticket.ts (hunk 12 above)
Comment (jfrerich) [reply src/forms/create_ticket.ts:185]: You cannot save a drop-down field in Zendesk without at least one option.  I'm not sure the best way to handle the check for that case.  If options are undefined, I suppose that I could just not show that field?

Comment (jfrerich) [reply src/forms/create_ticket.ts]: Ok. I added a check for defined options and if they don't exist, the field is added, but without any options. Again, this case should never occur because of the zendesk restrictions when created a drop box.

File: src/forms/create_ticket.ts [hunk 30]
```diff
@@ -170,14 +200,14 @@ class FormFields extends BaseFormFields {
         });
     }
 
-    isSystemField(field: Users.Fields.UserField): boolean {
-        return (SystemFields.includes(String(field.type)) || field.system_field_options);
+    isSystemField(field: ZDUserField): boolean {
+        return (SystemFields.includes(String(field.type)) || Boolean(field.system_field_options));
```
Comment (mickmister) [inline src/forms/create_ticket.ts:204]: What's the need to cast via `String()` here?

File: src/forms/subscriptions.ts [hunk 31]
```diff
@@ -83,18 +97,20 @@ class FormFields extends BaseFormFields {
...
+            results.push(this.addChannelTrigger(trigger));
         }
+        await Promise.all(results);
     }
 
-    async addChannelTrigger(trigger: any): Promise<void> {
+    async addChannelTrigger(trigger: ZDTrigger): Promise<void> {
         const parsedTitle = parseTriggerTitle(trigger.title);
         const channelID = parsedTitle.channelID;
         const channel = await this.mmClient.getChannel(channelID);
```
Comment (mickmister) [inline src/forms/subscriptions.ts:116]: Could we do some caching here during the request, so we don't fetch the same channel more than once?
